### PR TITLE
Added CSV branch to SoftJets and prescale weights for HT triggers

### PR DESCRIPTION
**Pull request by sbein**
Added b-tagging information to the soft jets, so they can be used for rebalance and smear.

### CODE REVIEW

**Comment by kpedro88** [comment]
@sbein did you mean to add a bunch of commits to this branch?

**Comment by sbein** [comment]
Yes, because I added b-tagging to the soft jets in one commit, then later made more updates and had to commit them separately. One of the commits was because some files in the Production/test directory got added that shouldn't have when I re-initialized my branch, and I needed to delete them and do a re-commit.

**Comment by kpedro88** [comment]
The first two new commits seem to be a different topic. I would prefer to have them in a separate PR, but they should at least be described in a message in this PR.

The second two new commits are not acceptable. Calling `git add` and then `git rm` does not undo the presence of the unwanted files in the history, which inflates the size of the repository. Please remove these commits from the branch:

```
git reset --soft HEAD~2
git push -f origin
```

or:

```
git rebase -i HEAD~3
(remove unwanted commits from list, then continue)
git push -f origin
```

**Comment by sbein** [comment]
Ok, I did 
git reset --soft HEAD~2
git push -f origin
So maybe that works now?

**Comment by kpedro88** [comment]
It doesn't seem to have updated the PR. Are you sure you pushed the branch sbein:Run2, and you hadn't added any commits locally since you pushed before?

**Comment by sbein** [comment]
Ah, I was in the origin branch. Now it looks like there are only 3 commits in the pull request. Let me know if that doesn't work.

**Comment by kpedro88** [comment]
Okay, that appears to work now.

A few review comments on the new additions:
1) As requested, please provide some description of what they do.
2) Suggestions for more standardized branch names:
- MHTOnline
- HTOnline
- PrescaleWeightHT  

3) Please remove commented out code from the new producer.

**Comment by sbein** [comment]
Ok I made the requested changes and documented the branch meanings in a new commit message.

**Comment by kpedro88** [comment]
Thanks, I'll try to test this tomorrow.

**Comment by kpedro88** [comment]
Is there some way you can put all the hardcoded numbers into python or a text file that the producer can load? I guess this the output of a specific training for the MET110 trigger. Is the structure of the network also contingent on the trigger or does it have more general applicability?

**Comment by sbein** [comment]
As per the discussion today, I have moved the code for the BNN efficiency to the twiki
https://twiki.cern.ch/twiki/bin/view/CMS/RA2b13TeVProduction#Trigger
And removed the last two commits in this pull request

File: Utils/src/PrescaleWeightProducer.cc [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+  virtual void produce(edm::Event&, const edm::EventSetup&);
+  virtual void beginJob();
+  virtual void endJob();
+  virtual void beginRun(edm::Run&, edm::EventSetup const&);
+  const double _startWeight;
+  const double _LumiScale;
+  const double _PrescaleCut;
+  const bool _PFHTWeights;
+  edm::InputTag _weightName;
+  string _processName;
```
Comment (kpedro88) [inline Utils/src/PrescaleWeightProducer.cc]: This variable appears to be initialized but never used. If it's not needed, please remove from the cc and py files

File: Utils/src/PrescaleWeightProducer.cc [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+
+  edm::EDGetTokenT<edm::TriggerResults> _triggerBits;
+  edm::EDGetTokenT<pat::TriggerObjectStandAloneCollection> _triggerObjects;
+  edm::EDGetTokenT<pat::PackedTriggerPrescales> _triggerPrescales;
+};
+
+PrescaleWeightProducer::PrescaleWeightProducer(const edm::ParameterSet& iConfig) :
+  _startWeight(iConfig.getParameter<double> ("weight")), _LumiScale(iConfig.getParameter<double> ("LumiScale")), _PrescaleCut(iConfig.getParameter<double> ("PrescaleCut")), _PFHTWeights(iConfig.getParameter<bool> ("PFHTWeights")), _weightName(iConfig.getParameter<edm::InputTag> ("weightName")), _processName(iConfig.getParameter<string> ("HLTProcess")), _triggerBits(consumes<edm::TriggerResults>(iConfig.getParameter<edm::InputTag>("bits"))), _triggerObjects(consumes<pat::TriggerObjectStandAloneCollection>(iConfig.getParameter<edm::InputTag>("objects"))), _triggerPrescales(consumes<pat::PackedTriggerPrescales>(iConfig.getParameter<edm::InputTag>("prescales")))
```
Comment (kpedro88) [inline Utils/src/PrescaleWeightProducer.cc]: Can you split this initialization list into a few lines so it doesn't stretch toward infinity?

**Comment by kpedro88** [comment]
When I run this on a file from `/QCD_HT200to300_TuneCUETP8M1_13TeV-madgraphMLM-pythia8/RunIISpring16MiniAODv1-PUSpring16_80X_mcRun2_asymptotic_2016_v3-v1/MINIAODSIM`, the branches `HTOnline`, `MHTOnline`, and `PrescaleWeightHT` are all 0. I guess this is because the trigger simulation was bad in the ICHEP MC?

What sample(s) did you use to test this code?

**Comment by sbein** [comment]
I ran it on a data MET sample using the command:
cmsRun runMakeTreeFromMiniAOD_cfg.py scenario=2016EF inputFilesConfig=Run2016E-PromptReco-v2.MET nstart=2 nfiles=10 outfile=MET16E numevents=100

I don't intend to use these branches in the MC, so I've conditioned their storage based on the geninfo flag. I'm pushing this now.

**Comment by kpedro88** [comment]
please address the code review comments on this PR

File: Utils/python/prescaleweightproducer_cfi.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
+import FWCore.ParameterSet.Config as cms
+
+prescaleweightProducer = cms.EDProducer('PrescaleWeightProducer',
+
+   HLTProcess = cms.string("HLT"),
+
+   PFHTWeights = cms.bool( True ),
```
Comment (kpedro88) [inline Utils/python/prescaleweightproducer_cfi.py]: it looks like this param could now be deleted as well

Comment (sbein) [reply Utils/python/prescaleweightproducer_cfi.py]: Ok, I took care of it.

**Comment by kpedro88** [comment]
Testing now...

**Comment by kpedro88** [comment]
I ran the same test that you specified, and the results make sense (the few events which pass PFHT800 have PrescaleWeightHT=1).

**Comment by sbein** [comment]
Yep, it's nice to see the that trigger has not been prescaled...yet

**Pull request merged**